tools/qemu-kvm-patch/kvm-qemu-2.4.0.1-libvmi.patch

```diff
@@ -0,0 +1,372 @@
++    free(pargs->path);
++    free(pargs);
++error_exit:
++    return NULL;
++}
++
++void
++qmp_pmemaccess (const char *path, Error **errp)
++{
++    pthread_t thread;
++    sigset_t set, oldset;
++    struct pmemaccess_args *pargs;
++
++    // create the args struct
++    pargs = (struct pmemaccess_args *) malloc(sizeof(struct pmemaccess_args));
++    pargs->errp = errp;
++    // create a copy of path that we can safely use
++    pargs->path = malloc(strlen(path) + 1);
++    memcpy(pargs->path, path, strlen(path) + 1);
++	
++    // start the thread
++    sigfillset(&set);
++    pthread_sigmask(SIG_SETMASK, &set, &oldset);
++    pthread_create(&thread, NULL, memory_access_thread, pargs);
++    pthread_sigmask(SIG_SETMASK, &oldset, NULL);
++
++}
+diff -Naur qemu-2.4.0.1/memory-access.h qemu-2.4.0.1-valerio/memory-access.h
+--- qemu-2.4.0.1/memory-access.h	1969-12-31 17:00:00.000000000 -0700
++++ qemu-2.4.0.1-valerio/memory-access.h	2015-10-12 10:15:52.587843591 -0600
+@@ -0,0 +1,21 @@
++/*
++ * Mount guest physical memory using FUSE.
++ *
++ * Author: Valerio G. Aimale <valerio@aimale.com>
++ */
++
++#ifndef MEMORY_ACCESS_H
++#define MEMORY_ACCESS_H
++
++#include "qapi-types.h"
++#include "qapi/qmp/qdict.h"
++#include "qapi/error.h"
++
++void qmp_pmemaccess (const char *path, Error **errp);
++
++struct pmemaccess_args {
++	char *path;
++	Error **errp;
++};
++
++#endif /* MEMORY_ACCESS_H */
+diff -Naur qemu-2.4.0.1/qapi-schema.json qemu-2.4.0.1-valerio/qapi-schema.json
+--- qemu-2.4.0.1/qapi-schema.json	2015-09-22 16:00:51.000000000 -0600
++++ qemu-2.4.0.1-valerio/qapi-schema.json	2015-10-12 10:28:33.959819998 -0600
+@@ -1389,6 +1389,34 @@
+   'data': {'val': 'int', 'size': 'int', 'filename': 'str'} }
+ 
+ ##
++# @pmemaccess:
++#
++# Open A UNIX Socket access to physical memory
++#
++# @path: the name of the UNIX socket pipe
++#
++# Returns: Nothing on success
++#
++# Since: 2.4.0.1
++#
++# Notes: Derived from previously existing patches. When command
++# succeeds connect to the open socket. Write a binary structure to 
++# the socket as:
++# 
++# struct request {
++#     uint8_t type;   // 0 quit, 1 read, 2 write, ... rest reserved
++#     uint64_t address;   // address to read from OR write to
++#     uint64_t length;    // number of bytes to read OR write
++# };
++# 
++# If it is a read operation, Qemu will return lenght+1 bytes. Read lenght+1
++# bytes. the last byte will be a 1 for success, or a 0 for failure.
++#  
++##
++{ 'command': 'pmemaccess',
++  'data': {'path': 'str'} }
++
++##
+ # @cont:
+ #
+ # Resume guest VCPU execution.
+diff -Naur qemu-2.4.0.1/qmp-commands.hx qemu-2.4.0.1-valerio/qmp-commands.hx
+--- qemu-2.4.0.1/qmp-commands.hx	2015-09-22 16:00:51.000000000 -0600
++++ qemu-2.4.0.1-valerio/qmp-commands.hx	2015-10-12 10:37:43.791802959 -0600
+@@ -467,6 +467,29 @@
+ EQMP
+ 
+     {
++        .name       = "pmemaccess",
++        .args_type  = "path:s",
++        .mhandler.cmd_new = qmp_marshal_input_pmemaccess,
```

Comment:
For anyone wondering about QEMU 2.5,

When you try to build this against QEMU 2.5.1 it'll end up looking for qmp_marshal_pmemaccess instead.

I'm guessing they derive the names from the schema automatically or something.